forte/data/vocabulary.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import List, Tuple, Dict, Union, Hashable, Iterable
+
+
+class Vocabulary:
+    '''This class maps element to representation. Element
+    could be any hashable type and there are two types of
+    representations, namely, "indexing" and "one-hot". There
+    are two special types of element, namely PAD_ELEMENT
+    and UNK_ELEMENT.
+    For "indexing" vocabulary,
+        Element:  <PAD>  ele1   ele2   ele3  ...
+        Id:         0      1      2      3   ...
+        Repr:       0      1      2      3   ...
+    For "one-hot" vocabulary,
+        Element:  <PAD>  ele1   ele2   ele3  ...
+        Id:        -1      0      1      2   ...
+        Repr:      [0,    [1,    [0,    [0,  ...
+                    0,     0,     1,     0,  ...
+                    0,     0,     0,     1,  ...
+                    0,     0,     0,     0,  ...
+                    ...]   ...]   ...]   ...]
+    If vocabulary uses UNK_ELEMENT, the first element,
+    "ele1" will be UNK_ELEMENT and any other elements
+    that cannot be found in the current vocabulary will
+    be mapped to the UNK_ELEMENT. Otherwise, UNK_ELEMENT
+    is not used. Error will occur when querying unknown
+    element in the vocabulary.
+    '''
+    PAD_ELEMENT = "<PAD>"
+    UNK_ELEMENT = "<UNK>"
```

Comment:
Is it complicated to support more special tokens? For example, can the user pass one a set of "<SPECIAL_1>", "<SPECIAL_2>"

Reply:
I have a function outside the vocabulary and in the extractor to add some predefined element into the vocabulary. And the user can pass special tokens as they want. 

Since the vocabulary here is for all kinds of element (token, word, label, seq-tag), I don't want to make assumption that there will be other special element other than <PAD> and <UNK>. 

These special elements are supported, but it is done by caller outside the vocabulary (possibly extractor or actual user) to call the `add` function to add these special token. The id they use will be the order they pass in.